drivers/adc/adc_ad4114.c

```diff
@@ -0,0 +1,462 @@
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct adc_ad4114_data *data = CONTAINER_OF(ctx, struct adc_ad4114_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_ad4114x_validate_buffer_size(const struct device *dev,
+					    const struct adc_sequence *sequence)
+{
+	uint8_t channels;
+	size_t needed;
+
+	channels = POPCOUNT(sequence->channels);
+	needed = channels * sizeof(uint32_t);
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_ad4114_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_ad4114_data *data = dev->data;
+	const struct adc_ad4114_config *config = dev->config;
+	int ret;
+	uint8_t write_reg[2];
+	uint8_t status;
+
+	ret = adc_ad4114x_validate_buffer_size(dev, sequence);
+	if (ret < 0) {
+		LOG_ERR("insufficient buffer size");
+		return ret;
+	}
+
+	data->channels_cfg = sequence->channels;
+	for (uint32_t i = 0U; i < AD4114_CHAN_NUMBER; i++) {
+		if ((BIT(i) & sequence->channels) != 0) {
+			write_reg[0] = 0x80 | (uint8_t)((config->map_input[i] >> 8) & 0xFF);
+			write_reg[1] = (uint8_t)(config->map_input[i] & 0xFF);
+			LOG_DBG("Enable channel %d with mapping %X %X, raw %X", i, write_reg[0],
+				write_reg[1], config->map_input[i]);
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		} else {
+			LOG_DBG("Disable channel %d", i);
+			write_reg[0] = 0x0;
+			write_reg[1] = 0x0;
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		}
+	}
+
+	/* Configure the buffer */
+	data->buffer = sequence->buffer;
+
+	while ((status & 0x80) != 0x80) {
+		/* Wait for acquiistion start */
+		ad4114_read_reg(dev, AD4114_STATUS_REG, &status, 1);
+		k_usleep(10);
+	}
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static void adc_ad4114_acquisition_thread(struct adc_ad4114_data *data)
+{
+	uint8_t value[4] = {0};
+	uint32_t buffer_values[AD4114_CHAN_NUMBER];
+	bool is_ended = false;
+
+	while (true) {
+		k_sem_take(&data->sem, K_FOREVER);
+
+		while (data->channels != 0) {
+			ad4114_read_reg(data->dev, AD4114_DATA_REG, value, 4);
+			/* Check the read channel */
+			if ((value[3] & 0xF0) != 0) {
+				LOG_DBG("Error read on :  %X ", value[3]);
+			} else {
+				LOG_DBG("Success read on %d: value  %X ", value[3],
+					(value[2] << 16 | value[1] << 8 | value[0]));
+				/* success read, store it */
+				buffer_values[value[3]] =
+					(value[0] << 16 | value[1] << 8 | value[2]);
+				WRITE_BIT(data->channels, value[3], 0);
+				/* Disable the channel after read success */
+				uint8_t write_reg[2] = {0};
+
+				ad4114_write_reg(data->dev, AD4114_CHANNEL_0_REG + value[3],
+						 write_reg, 2);
+			}
+			if (data->channels == 0) {
+				is_ended = true;
+			}
+			k_usleep(10);
```

Comment:
Document with a comment why you chose this sleep duration. Please do this for all of the sleeps

Reply:
Comments added